OpenRA.Mods.Common/Traits/Buildings/ProximityLinkHost.cs

```diff
@@ -0,0 +1,250 @@
+		{
+			if ((clientActor.CenterPosition - LinkPosition).HorizontalLengthSquared > Info.Range.LengthSquared)
+			{
+				moveCooldownHelper.NotifyMoveQueued();
+
+				// TODO: MoveWithinRange doesn't support offsets.
+				// TODO: MoveWithinRange considers the whole footprint instead of a point on the actor.
+				moveToLinkHostActivity.QueueChild(clientActor.Trait<IMove>().MoveWithinRange(Target.FromActor(self), Info.Range));
+				return true;
+			}
+
+			return false;
+		}
+
+		public virtual void QueueLinkActivity(Activity moveToLinkHostActivity, Actor self, Actor clientActor, LinkClientManager client)
+		{
+			if (Info.OccupyClient)
+			{
+				if (moveToLinkHostActivity == null)
+					clientActor.QueueActivity(new ProximityLinkSequence(client, self, this, Info.Range.Length));
+				else
+					moveToLinkHostActivity.QueueChild(new ProximityLinkSequence(client, self, this, Info.Range.Length));
+			}
+			else
+			{
+				// Make sure OnLinkStarted is only called once.
+				if (!WaitingClients.Any(p => p.Client.Trait == client) && !LinkedClients.Any(p => p.Trait == client))
+					OnLinkStarted(self, clientActor, client);
+			}
+		}
+
+		#endregion
+
+		void ITick.Tick(Actor self)
+		{
+			Tick(self);
+		}
+
+		bool ClientAliveAndInRange(Actor clientActor) => clientActor != null && !clientActor.IsDead && clientActor.IsInWorld
+			&& (clientActor.CenterPosition - LinkPosition).HorizontalLengthSquared <= Info.Range.LengthSquared;
+
+		protected virtual void Tick(Actor self)
+		{
+			if (Info.OccupyClient || IsTraitDisabled)
+				return;
+
+			// Track wait time manually.
+			for (var i = 0; i < WaitingClients.Count; i++)
+			{
+				if (WaitingClients[i].Time < self.World.WorldTick)
+				{
+					var client = WaitingClients[i].Client.Trait;
+					if (ClientAliveAndInRange(WaitingClients[i].Client.Actor) && client.CanLinkTo(self, this, false, true))
+						LinkStarted(self, WaitingClients[i].Client);
+					else
+						Unreserve(client);
+
+					WaitingClients.RemoveAt(i);
+					i--;
+				}
+			}
+
+			// Tick clients manually.
+			for (var i = 0; i < LinkedClients.Count; i++)
+			{
+				var clientActor = LinkedClients[i].Actor;
+				if (ClientAliveAndInRange(clientActor) && !LinkedClients[i].Trait.OnLinkTick(clientActor, self, this))
+					continue;
+
+				OnLinkCompleted(self, clientActor, LinkedClients[i].Trait);
+				i--;
+			}
+		}
+
+		protected virtual void LinkStarted(Actor self, TraitPair<LinkClientManager> client)
+		{
+			LinkedClients.Add(client);
+
+			foreach (var nd in client.Actor.TraitsImplementing<INotifyLinkClient>())
+				nd.Linked(client.Actor, self);
+
+			foreach (var ndh in notifyLinkHosts)
+				ndh.Linked(self, client.Actor);
+
+			client.Trait.OnLinkStarted(client.Actor, self, this);
+		}
+
+		public virtual void CancelLink(Actor self)
+		{
+			// Cancelling will be handled in the RemoteLinkSequence activity.
```

Comment:
Shouldn't it be `ProximityLinkSequence`?:

```suggestion
			// Cancelling will be handled in the ProximityLinkSequence activity.
```

Reply:
fixed in #20665